### PR TITLE
Set property

### PR DESCRIPTION
**Pull request by ArArgon**
- [ ] Investigate set prop cache miss
- [ ] Implement delete prop (basically disable IC when deletion happens)

### CODE REVIEW

File: core/src/avm2/activation.rs [hunk 1]
```diff
@@ -1484,7 +1496,7 @@ impl<'a, 'gc> Activation<'a, 'gc> {
                         name_value.as_object().unwrap(),
                         self.context.gc_context,
                     );
-
+                    object.remove_shape_id(self.context.gc_context);
```
Comment (xyjixyjixyji) [inline core/src/avm2/activation.rs]: Please double check that does slot id change or not upon delete property bytecode.

File: core/src/avm2/object.rs [hunk 2]
```diff
@@ -413,10 +454,18 @@ pub trait TObject<'gc>: 'gc + Collect + Debug + Into<Object<'gc>> + Clone + Copy
...
-                let base = self.base();
+        // Try ic
+        let base = self.base();
+        if let Some(shape_id) = base.shape_id() {
+            if self.try_set_property_by_shape(activation, multiname, value, shape_id, ic)? {
+                return Ok(());
+            }
+        }
+
+        match self.vtable().get_trait_with_ns(multiname) {
```
Comment (xyjixyjixyji) [inline core/src/avm2/object.rs:467]: i don't think we need to use `get_trait_with-ns`, even the original solution is not using it

File: core/src/avm2/object.rs [hunk 3]
```diff
@@ -426,11 +475,15 @@ pub trait TObject<'gc>: 'gc + Collect + Debug + Into<Object<'gc>> + Clone + Copy
...
-                            multiname.namespace_set().to_vec(),
+                            vec![ns],
                             PropertyType::Property(Property::Slot { slot_id }),
                         ),
                     );
-                    base.set_shape_id(mc, new_shape_id);
+                    if new_shape_id != id {
+                        base.set_shape_id(mc, new_shape_id);
+                    } else {
+                        // update ic
```
Comment (xyjixyjixyji) [inline core/src/avm2/object.rs]: what is this else branch for

File: core/src/avm2/object.rs [hunk 4]
```diff
@@ -426,11 +475,15 @@ pub trait TObject<'gc>: 'gc + Collect + Debug + Into<Object<'gc>> + Clone + Copy
                         id,
                         PropertyInfo::new(
                             multiname.local_name().unwrap(),
-                            multiname.namespace_set().to_vec(),
+                            vec![ns],
```
Comment (xyjixyjixyji) [inline core/src/avm2/object.rs]: if so, we better compare the namespace in the `add_property`.

For example, if `ns1 <= ns2`, it should not be treated as a new property in the shape

File: core/src/avm2/object.rs [hunk 5]
```diff
@@ -1206,6 +1259,11 @@ pub trait TObject<'gc>: 'gc + Collect + Debug + Into<Object<'gc>> + Clone + Copy
         base.set_shape_id(mc, shape_id);
     }
 
+    #[no_dynamic]
+    fn remove_shape_id(&self, mc: &Mutation<'gc>) {
```
Comment (xyjixyjixyji) [inline core/src/avm2/object.rs]: ditto, not sure this is needed or not

File: core/src/avm2/object/script_object.rs [hunk 6]
```diff
@@ -335,6 +334,8 @@ impl<'gc> ScriptObjectWrapper<'gc> {
         }
         if let Some(name) = multiname.local_name() {
             let key = maybe_int_property(name);
+
+            self.remove_shape_id(mc);
```
Comment (xyjixyjixyji) [inline core/src/avm2/object/script_object.rs]: ditto

File: core/src/avm2/object/script_object.rs [hunk 7]
```diff
@@ -482,6 +482,10 @@ impl<'gc> ScriptObjectWrapper<'gc> {
         unlock!(Gc::write(mc, self.0), ScriptObjectData, shape_id).set(Some(shape_id));
     }
 
+    pub fn remove_shape_id(&self, mc: &Mutation<'gc>) {
```
Comment (xyjixyjixyji) [inline core/src/avm2/object/script_object.rs]: ditto

File: core/src/avm2/object.rs (hunk 4 above)
Comment (xyjixyjixyji) [reply core/src/avm2/object.rs]: But i think this could be acceptable, depends on what you think

**Review by xyjixyjixyji — COMMENTED** [review]
Comments above other parts LGTM \cc @ArArgon

File: core/src/avm2/object.rs (hunk 5 above)
Comment (ArArgon) [reply core/src/avm2/object.rs]: Shape is now preserved.

**Pull request merged**